app/src/main/java/com/aletheiaware/perspective/android/WorldAdapter.java

```diff
@@ -102,31 +109,77 @@ public void onClick(View v) {
     }
 
     @Override
-    public void onBindViewHolder(@NonNull WorldViewHolder holder, int position) {
-        if (names.isEmpty()) {
-            holder.setEmptyView();
+    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
+        if (position == 0) {
+            StarCountHolder sch = (StarCountHolder) holder;
+            int earned = 0;
+            int max = 0;
+            for (String name : names) {
```

Comment:
When a world is added to this adapter it calls notifyDataSetChanged which tells the recyclerview to redraw. When the recyclerview redraws the header and binds it to the view it will calculate the stars earned and update the text view.

When the next world is added the above is repeated, and since multiple worlds are added the header is updated multiple times which gives a nice update animation.

It would be possible to keep a global count of earned stars, but the animation would have to be explicitly generated.

Reply:
Separating UI from business logic is a good thing, IMHO.

Reply:
In principle yes, and I think this already has good MVC separation;
- Controller - WorldSelectActivity adds worlds to the Adapter
- Model - Adapter determines how to display each world, and a summary of player progress
- View - RecyclerView renders ViewHolders from Adapter